src/controllers/healthcheck/healthcheck.controller.js

```diff
@@ -0,0 +1,192 @@
+
+/**
+ *
+ * @param {string[]} walletIds
+ * @param {boolean} includeFullnode
+ * @param {boolean} includeTxMiningService
+ * @returns {Promise<{checks: Object, httpStatus: number, status: string}>}
+ * @private
+ * @description Returns the health checks for the given wallet ids. The fullnode and the
+ * tx-mining-service are optionally included in the checks, depending on the parameters.
+ *
+ * Also returns the http status code and the overall status of the health check.
+ *
+ * Returned object format:
+ *     {
+ *       httpStatus: 200,
+ *       status: 'pass',
+ *       checks: {
+ *        'Wallet health_wallet': [
+ *          {
+ *            componentName: 'Wallet health_wallet',
+ *            componentType: 'internal',
+ *            status: 'fail',
+ *            output: 'Wallet is not ready. Current state: Syncing',
+ *            time: expect.any(String),
+ *          },
+ *        ],
+ *        'Fullnode http://fakehost:8083/v1a/': [
+ *          {
+ *            componentName: 'Fullnode http://fakehost:8083/v1a/',
+ *            componentType: 'fullnode',
+ *            status: 'pass',
+ *            output: 'Fullnode is responding',
+ *            time: expect.any(String),
+ *          },
+ *        ],
+ *        'TxMiningService http://fake.txmining:8084/': [
+ *          {
+ *            componentName: 'TxMiningService http://fake.txmining:8084/',
+ *            componentType: 'service',
+ *            status: 'pass',
+ *            output: 'Tx Mining Service is healthy',
+ *            time: expect.any(String),
+ *          },
+ *        ],
+ *      }
+ *    }
+ *
+ */
+async function getWalletsHealthChecks(
+  walletIds,
+  includeFullnode = false,
+  includeTxMiningService = false
+) {
+  const promises = Array.from(walletIds).map(
+    walletId => healthService.getWalletHealth(initializedWallets.get(walletId), walletId)
+  );
+
+  if (includeFullnode) {
+    promises.push(healthService.getFullnodeHealth());
+  }
+  if (includeTxMiningService) {
+    promises.push(healthService.getTxMiningServiceHealth());
+  }
+
+  const resolvedPromises = await Promise.all(promises);
+
+  const checks = {};
+
+  for (const healthData of resolvedPromises) {
+    // We use an array as the value to stick to our current format,
+    // which allows us to add more checks to a component if needed
+    checks[healthData.componentName] = [healthData];
+  }
+
+  const httpStatus = resolvedPromises.every(healthData => healthData.status === 'pass') ? 200 : 503;
+  // If any of the checks failed, the status is fail. If all checks passed, the status is pass.
+  // Otherwise, the status is warn.
+  let status = httpStatus === 200 ? 'pass' : 'warn';
+  status = resolvedPromises.some(healthData => healthData.status === 'fail')
+    ? 'fail'
+    : status;
+
+  return { checks, httpStatus, status };
+}
+
+/**
+ * Controller for the /health endpoint that returns the health
+ * of the wallet-headless service, including all started wallets,
```

Comment:
I thought the wallet-headless service health endpoint would be something more direct and not depending on the wallet itself. I thought would be just an endpoint to check if the service is up.

Reply:
After some rounds of review in the overall design we headed towards a solution where the services' health depends on its components health.

There are some comments about this in https://github.com/HathorNetwork/internal-rfcs/pull/5#discussion_r1282053331 and https://github.com/HathorNetwork/internal-rfcs/pull/5#discussion_r1282218526, and I believe we discussed this in meetings as well.

For what we are trying to achieve, which is to have an overall visualization of the health of all our systems, I believe this is the way to go. If we only did a simple endpoint like you mentioned, we wouldn't have any feedback about problems in, say, the fullnode that is serving a wallet-headless, or the tx-mining-service, or the wallet itself.

But this is still open for discussion, of course, if you think there is a problem in this approach. We could also discuss which components should be included in the healthcheck. You could think that we shouldn't include the wallet, but we should include the fullnode, for instance. Changing this would be simple.

----

With that said, there is a pattern out there, used mostly in Kubernetes environments, which is to have both a `readiness` and a `liveness` healthcheck.

They seem to be close to what we are discussing here. You think we should have only a `liveness` healthcheck, which only checks if the service is alive. I've implemented a `readiness` healthcheck, which also checks if it's ready to operate (this is a rough description).

But, as I said above, I think the `readiness` one gives us much more feedback about the health of the whole service. I think we want to know whether the Wallet is in a "not ready" state for long periods of time, for instance. This could indicate a problem somewhere.